Controlling the background color of the active tab in an unfocused editor group

dmolin opened the pull request:
It's currently not possible to set the background color of the active tab in un unfocused editor group.
With this PR we can now control the background color for the active tab in an unfocused editor group through the "tab.unfocusedActiveBackground" setting (under ""workbench.colorCustomizations").

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=73848) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=73848) before we can accept your contribution.<br/><br/>:x: dmolin [sign now](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=73848)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vscode?pullRequest=73848) it.</sub>

dmolin commented:
Thanks for the feedback @bpasero, highly appreciated, but I'm not sure I understand the extent of it. The only usages I see of TAB_ACTIVE_BACKGROUND are in the file where I did my changes. I see it used in dragging/dropping the tabs (something that I see working just fine with my change) and those places are not affected by the color of the active tab when unfocused (since the tab is focused at that time).

There's a place in `tabsTitleControl.ts` where "gradients" are reset when hovering on the tabs (active and inactive ones) but I don't think that's relevant, since the new setting is only supposed to determine the color of an "unfocused" active tab...

bpasero commented:
@dmolin this applies when the tabs are configured to shrink (as you can see from the `tab.sizing-shrink` CSS rule via the setting `"workbench.editor.tabSizing": "shrink"`) so I would think this also applies to your new color. Can you check?

dmolin commented:
Yeah you're right! 
I had to fight a bit with it (the unfocused style was overriding the focused one) but then I found what the issue was and fixed it: The current style for the active tab background was not taking into account the tab container state (`.active` vs `.inactive`). 
I added it to the selector and then I added the style for the unfocused (`.inactive`) container group. Everything seems to work perfectly (I tested it with both fit and shrink tab modes).

In the screenshot you can see how it looks in shrink mode.
![inactive-group-active-tab](https://user-images.githubusercontent.com/861513/57971390-4cf13c00-7985-11e9-9a98-a431727dd316.png)

I just pushed my changes. :smile:

dmolin commented:
I made the last few tweaks and tested it. Ready for review.

bpasero commented on src/vs/workbench/common/theme.ts:
```diff
@@ -138,6 +138,11 @@ export const TAB_UNFOCUSED_INACTIVE_FOREGROUND = registerColor('tab.unfocusedIna
 	hc: Color.white
 }, nls.localize('tabUnfocusedInactiveForeground', "Inactive tab foreground color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."));
 
+export const TAB_UNFOCUSED_ACTIVE_BACKGROUND = registerColor('tab.unfocusedActiveBackground', {
```
@dmolin can we do 2 changes here:
* the default colors (for dark, light, hc) should really be `TAB_ACTIVE_BACKGROUND` to preserve todays behaviour
* move this right below the definition of `TAB_ACTIVE_BACKGROUND`

dmolin replied:
done in commit 82f5ef9

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -1292,6 +1292,22 @@ registerThemingParticipant((theme: ITheme, collector: ICssStyleCollector) => {
 		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
 			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
 			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
+			collector.addRule(`
```
@dmolin I feel like there is room for optimization here because in 99% of the times, the 2 colors will be the same (tab active background, tab unfocused active background). Can we rewrite this such as we use the old CSS rule we used to have if both colors are the same and only split them up otherwise?

dmolin replied:
uhm.. I'm not sure here.
I actually followed the way the code is currently laid out (the same exact patter is already used for `tabHoverBackground` and `tabUnfocusedHoverBackground`).

Moreover, now the 2 CSS selectors are substantially different (the one for the active background explicitely uses `group-container.active`, the other doesn't). So, they would never share the current CSS (since one editor group is active, the other is not). There are ways to improve by generating the rule via composition (a prefix string that is different for the 2 rules plus a common one extracted into a variable) but the things is that this is not the approach taken for the other rules in this code block; I tried to maintain style consistency here, so I laid out the rule in the same way the others have already been written

dmolin replied:
@bpasero what I mean is, to make it work without code redundancies I could do this:

```
		const makeTabActiveBackgroundSelector = (color: Color, colorDrag: Color, isContainerActive = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${isContainerActive ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${isContainerActive ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;

		// Adjust gradient for active tab background
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundSelector(adjustedColor, adjustedColorDrag, true));
		}

		// Adjust gradient for unfocused active tab background
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);
		if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundSelector(adjustedColor, adjustedColorDrag));
		}
```

But that would be quite a different way to proceed compared to what has been already done for the aforementioned `tabHoverBackground` and `tabUnfocusedHoverBackground` (unless we agree to apply the same change there too, for the sake of consistency)

Let me know what your thoughts are about this

bpasero replied:
@dmolin yeah I think even the other rules could be tweaked. after all less CSS rules is better and faster if possible.

My idea was the following: if both colors are defined and identical you can simply generate a rule `.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container >` and that will work for both active and inactive groups.

dmolin replied:
@bpasero This could implement the "let's define the CSS rule only once" approach you were referring to:

I'm not a super fan of nested `if` but it's the easiest way to avoid too long ifs statements in the following blocks.
The idea, as you suggested, is that if the 2 colors are defined and equals we apply a common rule, otherwise we define 2 distinct specific rules.  Regarding your comment below related to the 2 colors not being customized, I do believe the common rule still applies (so, no reason to add a `:not(.active)` pseudo selector to the rule), since the 2 variable would always have a value and then would fall within the "common" rule.

```
		// Adjust gradient for active tab background (focused and unfocused editor groups)
		const makeTabActiveBackgroundRule = (color: Color, colorDrag: Color, hasFocus = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${hasFocus ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${hasFocus ? ".active" : ""} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);

		// if both colors (focused and unfocused) are equal, a common rule applies to both, otherwise we apply indivitual rules for
		// the .active and inactive editor groups
		if (tabActiveBackground && tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground
			&& tabActiveBackground.equals(tabUnfocusedActiveBackground)) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag));
		} else {
			if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
				const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
				const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
				collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag, true))
			}

			// Adjust gradient for unfocused active tab background
			if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
				const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
				const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
				collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag))
			}
		}
```

dmolin replied:
@bpasero I'm not sure that defining a common rule, though, is the best way to go... if the 2 colors are equal (that can happen if they are not customized but ALSO if only the active one is in the settings, since the unfocused one by default gets the same value as the active one), then you would get the "common" rule applied; since that rule does not enforce the `.active` part of the selector, the tab would end up being overridden by the rule that normally applies to "inactive" tabs!
(see image, where you can see that the :after pseudo selector for the active tab as been "taken over" by the rule that normally applies to inactive tabs):

![common-rule](https://user-images.githubusercontent.com/861513/58384011-c7007100-7fd4-11e9-928c-ef68dc8857ec.png)

So, I still believe that having the 2 conditions as before is the best way to go, maybe with the "make...Rule" to avoid writing twice (most of) the same code, like so:

```
		// Adjust gradient for active tab background (focused and unfocused editor groups)
		const makeTabActiveBackgroundRule = (color: Color, colorDrag: Color, hasFocus = false) => `
				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container${hasFocus ? ".active" : ":not(.active)"} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${color}, transparent);
				}

				.monaco-workbench .part.editor > .content.dragged-over .editor-group-container${hasFocus ? ".active" : ":not(.active)"} > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
					background: linear-gradient(to left, ${colorDrag}, transparent);
				}
		`;
		
		// Adjust gradient for unfocused active tab background
		const tabActiveBackground = theme.getColor(TAB_ACTIVE_BACKGROUND);
		if (tabActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag, true))
		}

		// Adjust gradient for unfocused active tab background
		const tabUnfocusedActiveBackground = theme.getColor(TAB_UNFOCUSED_ACTIVE_BACKGROUND);
		if (tabUnfocusedActiveBackground && adjustedTabBackground && adjustedTabDragBackground) {
			const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
			const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
			collector.addRule(makeTabActiveBackgroundRule(adjustedColor, adjustedColorDrag))
		}
```

bpasero replied:
Yeah good point.

dmolin replied:
one question @bpasero : the pre-commit hook doesn't allow me to commit because of this `.editor-group-container${hasFocus ? ".active" : ""}`
It says that there are "unexternalized strings"... where am I supposed to put those constants?

bpasero replied:
@dmolin simply use single quotes instead of double quotes 👍

dmolin replied:
I pushed the latest commit, essentially with the solution mentioned in my last comment (applied also to the `tabHoverBackground` rule. See f763adc

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -1292,6 +1292,22 @@ registerThemingParticipant((theme: ITheme, collector: ICssStyleCollector) => {
+			const adjustedColor = tabUnfocusedActiveBackground.flatten(adjustedTabBackground);
+			const adjustedColorDrag = tabUnfocusedActiveBackground.flatten(adjustedTabDragBackground);
 			collector.addRule(`
 				.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container > .title .tabs-container > .tab.sizing-shrink.active:not(.dragged) > .tab-label::after {
```
@dmolin I just realised that for this specific rule we need to use `.monaco-workbench .part.editor > .content:not(.dragged-over) .editor-group-container:not(.active)` (note the extra `:not(.active)`) to ensure we only define it for inactive groups and not overwrite the style for active groups. Because it could be that no color for active group is defined and then this rule would still apply.

dmolin replied:
you're right, good catch! :+1:

dmolin replied:
Wait.. is this a thing? I mean, can it actually happen? from what I see, both colors are always defined: if not present in any `settings.json`, they simply get the same color (taken from the editor background color). In that case, with the rewrite that applies the CSS rule only once, they would fall into the "common" rule case (the one that doesn't even add ".active" to the css selector) and the ".active" rule would never be defined in that case

bpasero replied:
@dmolin yeah in this case you are right the color will always be defined

dmolin replied:
see latest commit f763adc

dmolin commented on src/vs/workbench/common/theme.ts:
```diff
@@ -30,6 +30,12 @@ export const TAB_ACTIVE_BACKGROUND = registerColor('tab.activeBackground', {
+	dark: TAB_ACTIVE_BACKGROUND,
+	light: TAB_ACTIVE_BACKGROUND,
+	hc: TAB_ACTIVE_BACKGROUND
+}, nls.localize('tabUnfocusedActiveBackground', "tab background color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."));
```
fixed in f763adc

bpasero commented:
@dmolin should I do a final review, I think I can merge this today

dmolin commented:
@bpasero yes, go for it :+1:

bpasero commented:
@dmolin is there a reason you chose to use `.editor-group-container${hasFocus ? '.active' : ''}` in the one case for hover and `.editor-group-container${hasFocus ? '.active' : ':not(.active)'}` in the other?

dmolin commented:
> @dmolin is there a reason you chose to use `.editor-group-container${hasFocus ? '.active' : ''}` in the one case for hover and `.editor-group-container${hasFocus ? '.active' : ':not(.active)'}` in the other?

I thought about adding `:not(.active)` also for the tabHover; Keep in mind, that code was alreay there before and was `not` using it. My concern is if that was made on purpose (so, better leave it as it was) or if it something we can touch.. I went with being conservative since I'm not sure of the full extent of the possible side effect if adding ":not(.active)" there.

Tbh, I tested it with ":not(.active)" and it seemed to work just fine but, again, I'm not super sure if adding the more specific selector might have impacts on something else...

bpasero commented:
Ok thanks I will change and probably align.

dmolin commented:
@bpasero as you prefer. If you want I can make the change in a second, just let me know :smile:

bpasero commented:
@dmolin thanks, I merged it in.
